groups/service.go

```diff
@@ -0,0 +1,434 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"reflect"
+
+	admin "google.golang.org/api/admin/directory/v1"
+	"google.golang.org/api/googleapi"
+	groupssettings "google.golang.org/api/groupssettings/v1"
+	"google.golang.org/api/option"
+)
+
+const (
+	OwnerRole   = "OWNER"
+	ManagerRole = "MANAGER"
+	MemberRole  = "MEMBER"
+)
+
+// AdminService provides functionality to perform high level
+// tasks using a AdminServiceClient.
+type AdminService interface {
+	AddOrUpdateGroupMembers(group GoogleGroup, role string, members []string) error
+	CreateOrUpdateGroupIfNescessary(group GoogleGroup) error
+	DeleteGroupsIfNecessary() error
+	RemoveOwnerOrManagersFromGroup(group GoogleGroup, members []string) error
+	RemoveMembersFromGroup(group GoogleGroup, members []string) error
+	GetClient() AdminServiceClient
+}
+
+// GroupService provides functionality to perform high level
+// tasks using a GroupServiceClient.
+type GroupService interface {
+	UpdateGroupSettings(group GoogleGroup) error
+	GetClient() GroupServiceClient
```

Comment:
For followup: can we remove this? Not clear to me why we would want something other than the group service (or the thing building it) to have access to the client